it/uk/gov/hmrc/agentsubscriptionfrontend/controllers/BusinessIdentificationControllerISpec.scala

```diff
@@ -771,6 +811,130 @@ trait BusinessIdentificationControllerISpec extends BaseISpec with SessionDataMi
     }
   }
 
+  "showUpdateBusinessAddressForm" should {
+    behave like anAgentAffinityGroupOnlyEndpoint(request => controller.showBusinessNameForm(request))
+
+    "display update business address form if the address is not des complaint" in {
+      implicit val request = authenticatedAs(subscribingCleanAgentWithoutEnrolments)
+      sessionStoreService.currentSession.initialDetails = Some(initialDetails.copy(businessAddress = businessAddress.copy(addressLine1 =  "Address line 1 &")))
+
+      val result = await(controller.showUpdateBusinessAddressForm(request))
+      result should containMessages("updateBusinessAddress.title",
+        "updateBusinessAddress.p1", "updateBusinessAddress.p2", "updateBusinessAddress.address_line_1.title",
+        "updateBusinessAddress.address_line_2.title", "updateBusinessAddress.address_line_3.title",
+        "updateBusinessAddress.address_line_4.title", "updateBusinessAddress.postcode.title", "updateBusinessAddress.continue")
+
+      val doc = Jsoup.parse(bodyOf(result))
+      doc.getElementById("addressLine1").`val` shouldBe "Address line 1 &"
+      doc.getElementById("addressLine2").`val` shouldBe businessAddress.addressLine2.get
+      doc.getElementById("addressLine3").`val` shouldBe businessAddress.addressLine3.get
+      doc.getElementById("addressLine4").`val` shouldBe businessAddress.addressLine4.get
+      doc.getElementById("postcode").`val` shouldBe businessAddress.postalCode.get
+
+      val form = doc.select("form").first()
+      form.attr("method") shouldBe "POST"
+      form.attr("action") shouldBe routes.BusinessIdentificationController.submitUpdateBusinessAddressForm().url
+    }
+
+    "redirect to the /business-type page if there is no InitialDetails in session because the user has returned to a bookmark" in {
+      implicit val request = authenticatedAs(subscribingCleanAgentWithoutEnrolments)
+
+      val result = await(controller.showUpdateBusinessAddressForm(request))
+
+      resultShouldBeSessionDataMissing(result)
+    }
+  }
+
+  "submitUpdateBusinessAddressForm" should {
+    behave like anAgentAffinityGroupOnlyEndpoint(request => controller.submitBusinessNameForm(request))
+
+    "update business address after submission" in {
+      implicit val request =
+        authenticatedAs(subscribingCleanAgentWithoutEnrolments).withFormUrlEncodedBody("addressLine1" -> "new addressline 1",
+          "addressLine2" -> "new addressline 2",
+          "addressLine3" -> "new addressline 3",
+          "addressLine4" -> "new addressline 4",
+          "postcode" -> "BB11BB")
+      sessionStoreService.currentSession.initialDetails = Some(initialDetails)
+
+      val result = await(controller.submitUpdateBusinessAddressForm(request))
+      status(result) shouldBe 303
+      redirectLocation(result).head shouldBe routes.SubscriptionController.showCheckAnswers().url
+
+      val updatedBusinessAddress = await(sessionStoreService.fetchInitialDetails).get.businessAddress
```

Comment:
before I've seen that the InitialDetails have a countryCode along with the address, how come you're not checking for that

Reply:
We dont have any validation for country codes and we are not allowing users to edit that. However, there DES wont allow invalid country codes in backend.

Reply:
be nice to see that when they submit they always have a valid countryCode regardless